### PR TITLE
EffectParameter Additions / Completion

### PR DESCRIPTION
**Pull request by mattleibow**
I have added some changes to the `GetValue` methods to bring them up to the features of XNA. I also added a series of unit tests to ensure that I didn't break anything else.

Mostly, the current methods just catered for a single case, such as `GetValueMatrix`, which always failed if the data wasn't a 4x4 matrix. Now it supports a scalar, 2x3, 3x2, 3x4,4x3 and 4x4. (the 2x2,1x2, 2x1 seem to be missing from the other methods, so I didn't add them, but the can be)

A breaking change to the existing methods is the inclusion of a `int count` parameter to some of the `GetValueArray` methods. This is what XNA does and is a very useful feature. This size controls the size of the resulting array. For some reason, only some of these methods had the count parameter, such as `GetValueMatrixArray`.  
The breaking methods are:
- `GetValueSingleArray`
- `GetValueVector2Array`
- `GetValueVector3Array`
- `GetValueVector4Array`

These changes are the other side of #4268, I didn't touch those methods so that there will be no conflicts when merging.

NOTE: I made a whoopee somewhere and ended up with two PRs :( [#4299] Closed that one.

### CODE REVIEW

File: MonoGame.Framework/Graphics/Effect/EffectParameter.cs [hunk 1]
```diff
@@ -199,32 +199,165 @@ public int[] GetValueInt32Array ()
```
Comment (nkast) [inline MonoGame.Framework/Graphics/Effect/EffectParameter.cs]: TLeave a line note

**Comment by tomspilman** [comment]
I'm looking into why this PR is failing now.  Give me a sec.

> For some reason, only some of these methods had the count parameter

I'm sure that was something missed in the move from XNA 3 to XNA 4.  Great that you added it.

**Comment by mattleibow** [comment]
@tomspilman Is this all working now?

**Comment by tomspilman** [comment]
@KonajuGames - This look good to you?

File: MonoGame.Framework/Graphics/Vertices/VertexBufferBinding.cs [hunk 2]
```diff
@@ -99,5 +99,10 @@ public VertexBufferBinding(VertexBuffer vertexBuffer, int vertexOffset, int inst
             _vertexOffset = vertexOffset;
             _instanceFrequency = instanceFrequency;
         }
+
+        public static implicit operator VertexBufferBinding(VertexBuffer vertexBuffer)
+        {
+            return new VertexBufferBinding(vertexBuffer);
+        }
```
Comment (tomspilman) [inline MonoGame.Framework/Graphics/Vertices/VertexBufferBinding.cs]: Just noticed this change....  what was the need for this?

Comment (mattleibow) [inline MonoGame.Framework/Graphics/Vertices/VertexBufferBinding.cs]: This exists in the XNA build. I noticed it while porting an XNA game

**Comment by mattleibow** [comment]
@tomspilman @KonajuGames Any more work needed here?

**Comment by tomspilman** [comment]
This PR needs conflicts fixed.  Anyone up for that?

If not i'll close it next week.

**Comment by tomspilman** [comment]
Closing this... we can always reopen it if someone wants to merge it with the latest.

**Pull request closed without merging**